util/timer.h

```diff
@@ -56,28 +56,27 @@ class Timer {
   // the function takes 1000us to run. If it starts at time [now]us, then it
   // finishes at [now]+1000us, 2nd run starting time will be at [now]+3000us.
   // repeat_every_us == 0 means do not repeat.
-  void Add(std::function<void()> fn,
-           const std::string& fn_name,
-           uint64_t start_after_us,
-           uint64_t repeat_every_us) {
+  bool Add(std::function<void()> fn, const std::string& fn_name,
+           uint64_t start_after_us, uint64_t repeat_every_us) {
+    InstrumentedMutexLock l(&mutex_);
     std::unique_ptr<FunctionInfo> fn_info(new FunctionInfo(
         std::move(fn), fn_name, clock_->NowMicros() + start_after_us,
         repeat_every_us));
-    {
-      InstrumentedMutexLock l(&mutex_);
-      auto it = map_.find(fn_name);
-      if (it == map_.end()) {
-        heap_.push(fn_info.get());
-        map_.emplace(std::make_pair(fn_name, std::move(fn_info)));
-      } else {
-        // If it already exists, overriding it.
-        it->second->fn = std::move(fn_info->fn);
-        it->second->valid = true;
-        it->second->next_run_time_us = clock_->NowMicros() + start_after_us;
-        it->second->repeat_every_us = repeat_every_us;
-      }
+    // the new task start time should never before the current task executing
+    // time, as the executing task can only be running if it's next_run_time_us
+    // is due (<= clock_->NowMicros()).
+    assert(!executing_task_ ||
+           fn_info->next_run_time_us >= heap_.top()->next_run_time_us);
+    auto it = map_.find(fn_name);
+    if (it == map_.end()) {
+      heap_.push(fn_info.get());
```

Comment:
We can use `try_emplace()`

Reply:
I guess you mean `map_.try_emplace()`, updated.

Reply:
Sorry, I mean something like
```
auto res = map_.try_emplace(fn_name, std::move(fn_info));
if (!res.second) {
  return false;
}
heap_.push(res.first.second.get());
```